### PR TITLE
Send channel_update messages to direct peers on private channels

### PR DESCRIPTION
**Pull request by TheBlueMatt**
If we are a public node and have a private channel, our
counterparty needs to know the fees which we will charge to forward
payments to them. Without sending them a channel_update, they have
no way to learn that information, resulting in the channel being
effectively useless for outbound-from-us payments.

This commit fixes our lack of channel_update messages to private
channel counterparties, ensuring we always send them a
channel_update after the channel funding is confirmed.

Still need to fix existing tests/add new tests for the new behavior.

### CODE REVIEW

File: lightning/src/ln/channelmanager.rs
Comment (devrandom) [inline lightning/src/ln/channelmanager.rs:1546]: why send a channel update to the counterparty if the *counterparty* policy was violated?

File: lightning/src/ln/channelmanager.rs
Comment (devrandom) [inline lightning/src/ln/channelmanager.rs:1566]: nit: typos in last two lines

File: lightning/src/ln/channelmanager.rs
Comment (devrandom) [inline lightning/src/ln/channelmanager.rs:2373]: Don't we also want to send a unicast in this case if the channel is private?

File: lightning/src/ln/channelmanager.rs
Comment (devrandom) [inline lightning/src/ln/channelmanager.rs:950]: This is gated by the `if let Some(msg) = $funding_locked {` above.  Does it make sense to always send a channel update on reconnect, just in case we updated our policy while the channel was disconnected?

**Comment by devrandom** [comment]
Tested - correctly sends a `channel_update` right after `funding_locked`

File: lightning/src/ln/channelmanager.rs
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:1546]: Are you referring to the `amt_to_forward` value? It looks like the `htlc_minimum_msat` field in the generated channel update is `chan.get_counterparty_htlc_minimum_msat()`.

File: lightning/src/ln/channelmanager.rs
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:2373]: Hmmm, I'd think it doesn't matter much - its not like our counterparty doesn't know that we're disconnected. I'm not sure what other nodes do, but we should correctly flag a channel as `!is_usable` in the `ChannelDetails` and thus not include it in invoices.

File: lightning/src/ln/channelmanager.rs
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:950]: Hmm, good point, I pushed the channel_update sending down into the macro callsites.

File: lightning/src/ln/channelmanager.rs
Comment (devrandom) [reply lightning/src/ln/channelmanager.rs:1541]: whoa, this got attached to the wrong line, sorry

Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:1541]: No, this is being sent back to the HTLC sender. Practically its a no-op because the HTLC-sender shouldn't really store info about a private channel, but the spec says we have to include a channel_update for this error type. In reality, the bug is on the recipient's end, they shouldnt have included the channel hint in the invoice to begin with.

Comment (devrandom) [reply lightning/src/ln/channelmanager.rs:1541]: ah, got it now.  thank you.

**Comment by TheBlueMatt** [comment]
Fixed tests after rebasing on #954.

File: lightning/src/ln/channelmanager.rs
Comment (devrandom) [inline lightning/src/ln/channelmanager.rs:3429]: shouldn't this also check that the channel is private?

Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:3429]: I don't think we need to, no - when reconnecting its good to make sure that our direct channel counterparty has an updated copy of our channel_update, whether its public or not.

File: lightning/src/ln/channelmanager.rs
Comment (ariard) [inline lightning/src/ln/channelmanager.rs:3991]: Maybe add for "with channel_update for _private_", underscoring the privacy of the chan

**Review by ariard — COMMENTED** [review]
Overall SGTM

File: lightning/src/ln/channel.rs
Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:3999]: Oops, indeed. Note that this is #954.

**Comment by TheBlueMatt** [comment]
Rebased now that  #948 is merged, note that this still depends on # #948.

**Comment by TheBlueMatt** [comment]
Rebased on #954 and latest upstream, squashing the commits in this PR and fixing conflicts, also rebased on #972 and added a new commit which extends the test from it and adds a new bugfix.

**Comment by TheBlueMatt** [comment]
Rebased after all dependent PRs were merged.

**Comment by TheBlueMatt** [comment]
Fixed bench + fuzz failures due to the new messages which they weren't expecting.

File: lightning/src/ln/channelmanager.rs
Comment (valentinewallace) [inline lightning/src/ln/channelmanager.rs:4019]: Don't quite get why we don't push a `SendChannelUpdate` event when `announcement_sigs.is_some()`?

File: lightning/src/util/events.rs
Comment (valentinewallace) [inline lightning/src/util/events.rs:395]: nit: prefer `UnicastChannelUpdate` slightly for consistency

File: lightning/src/util/events.rs
Comment (valentinewallace) [inline lightning/src/util/events.rs:393]: nit: rephrase to: In contrast to [Self::BroadcastChannelUpdate], this is used when the channel...

File: lightning/src/ln/channelmanager.rs
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:4019]: If we send a `channel_update` which we intend to have relayed, we have to send a `channel_announcement` first. If we're just sending it to our counterparty they shouldn't care, but if the channel is public we have to wait until announcement_signatures are exchanged so we can send a channel_announcement.

File: lightning/src/util/events.rs
Comment (TheBlueMatt) [reply lightning/src/util/events.rs:395]: Hmm, all our other message events are `Send`..., though? I could move it up to sit next to the rest of the send message variants?

Comment (valentinewallace) [reply lightning/src/util/events.rs:395]: Ah, true. Resolving

File: lightning/src/ln/channelmanager.rs
Comment (valentinewallace) [inline lightning/src/ln/channelmanager.rs:3412]: Might be nicer to make `ChannelFlags` into its own struct and have methods to check who's node_one, etc. Not necessary for this PR though

**Review by valentinewallace — APPROVED** [review]

File: lightning/src/ln/channelmanager.rs
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:3412]: Yea, that's probably a good idea for the `channel_state` in `Channel` too, but we can do it later.

File: lightning/src/ln/channelmanager.rs
Comment (devrandom) [inline lightning/src/ln/channelmanager.rs:2985]: is it appropriate to send this message if we haven't seen the locking block?

File: lightning/src/ln/channelmanager.rs
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:1594]: Hmm, yea, happy to do something if you have a specific suggestion but I'm not sure what you really could do there. At least the `broadcast`/`unicast` naming makes it pretty clear IMO, especially when you're shoving it in a `Broadcast`... event

File: lightning/src/ln/channelmanager.rs
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:2985]: Hmm, good point. The spec is mum on it (it only mentions that we have to wait until we *receive* a `funding_locked`) see https://github.com/lightningnetwork/lightning-rfc/blob/master/07-routing-gossip.md#requirements-3 so I think its fine. There isn't a ton of logical reason to wait, either, IMO, its just some fields the counterparty stores locally, so who cares when its received.

File: lightning/src/ln/channelmanager.rs
Comment (devrandom) [reply lightning/src/ln/channelmanager.rs:1594]: Could have two thin wrappers around `msg::ChannelUpdate` for unicast/broadcast, but I haven't looked in detail whether it really makes sense.

File: lightning/src/ln/channelmanager.rs
Comment (devrandom) [reply lightning/src/ln/channelmanager.rs:2985]: OK, something to keep in mind if we see interop issues in this area.

**Review by devrandom — APPROVED** [review]

File: lightning/src/ln/channelmanager.rs
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:1594]: I dunno, its very explicit at the callsite already, making it more explicit doesn't seem like it'll solve anything IMO.

**Comment by TheBlueMatt** [comment]
Squashed without diff, will merge after CI:
```
$ git diff-tree -U1 6cf22aa2 ba600db7
$
```

**Pull request merged**
